### PR TITLE
Add schema validation and default value filling for extraction results

### PR DESCRIPTION
**Pull request by mohamedmamdouh22**
---

🛡️ This PR adds robust schema validation and default value filling for LLM extraction results to handle malformed JSON responses and ensure data consistency. It introduces a new schema validator module that automatically fills missing required fields with appropriate default values based on JSON schema types.

<details>
<summary>🔍 <strong>Detailed Analysis</strong></summary>

### Key Changes
- **New Schema Validator Module**: Created `schema_validator.py` with comprehensive validation and default value filling logic
- **AI Extraction Enhancement**: Modified `ai_extract` function to validate extraction results against provided schemas
- **Navigation Goal Integration**: Added schema validation to `extract_information_for_navigation_goal` function
- **Error Handling**: Implemented graceful fallback to original results if validation fails

### Technical Implementation
```mermaid
flowchart TD
    A[LLM Extraction Result] --> B{Schema Provided?}
    B -->|No| C[Return Original Result]
    B -->|Yes| D[validate_and_fill_extraction_result]
    D --> E[fill_missing_fields]
    E --> F{Field Type?}
    F -->|Object| G[Process Properties]
    F -->|Array| H[Process Items]
    F -->|Primitive| I[Return Value]
    G --> J{Required Field Missing?}
    J -->|Yes| K[Add Default Value]
    J -->|No| L[Skip Optional Field]
    K --> M[Recursively Process]
    L --> M
    H --> N[Process Each Item]
    N --> M
    I --> M
    M --> O[Return Validated Result]
```

### Impact
- **Data Consistency**: Ensures extraction results always conform to expected schema structure
- **Error Resilience**: Handles malformed LLM responses gracefully without breaking workflows
- **Type Safety**: Provides appropriate default values based on JSON schema types (string="", number=0, boolean=false, etc.)
- **Backward Compatibility**: Falls back to original results if validation fails, maintaining existing functionality

</details>

_Created with [Palmier](https://www.palmier.io)_

----

> [!IMPORTANT]
> Adds schema validation and default value filling for LLM extraction results, ensuring data consistency and handling malformed JSON responses.
> 
>   - **Behavior**:
>     - Adds `validate_and_fill_extraction_result` to `real_skyvern_page_ai.py` and `handler.py` to validate and fill extraction results based on schemas.
>     - Falls back to original results if validation fails.
>   - **Schema Validator**:
>     - New `schema_validator.py` with `validate_and_fill_extraction_result`, `fill_missing_fields`, and `get_default_value_for_type` functions.
>     - Handles JSON schema types, filling missing fields with defaults (e.g., string="", number=0).
>   - **Integration**:
>     - Modifies `ai_extract` in `real_skyvern_page_ai.py` to use schema validation.
>     - Updates `extract_information_for_navigation_goal` in `handler.py` to include schema validation.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=Skyvern-AI%2Fskyvern&utm_source=github&utm_medium=referral)<sup> for 13d7fadd015a7c294e4ab18e18106ecbd39cc118. You can [customize](https://app.ellipsis.dev/Skyvern-AI/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

## Summary by CodeRabbit

* **New Features**
  * Schema-based validation and auto-completion for LLM extraction results; extraction outputs are post-processed against provided schemas.

* **Reliability**
  * Improved conformity and traceability of extraction outputs; new explicit error type surfaces schema validation failures with detailed info.

* **Tests**
  * Comprehensive unit tests covering defaults, nested fills, nullable cases, edge cases, and validation flows.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

---

> [!NOTE]
> Adds JSON Schema-based validation/default-filling of extraction outputs and integrates it into extraction flows, with a new exception and tests.
> 
> - **Extraction/LLM**:
>   - **Schema validation & defaulting**: New `skyvern/forge/sdk/api/llm/schema_validator.py` to validate results, fill missing required fields, and report errors (`validate_and_fill_extraction_result`, `fill_missing_fields`, `validate_schema`, `validate_data_against_schema`).
>   - **Integration**: Apply validator in `real_skyvern_page_ai.ai_extract` and `webeye/actions/handler.extract_information_for_navigation_goal` when a `schema` is provided.
> - **Errors**:
>   - Add `InvalidSchemaError` in `skyvern/exceptions.py` for invalid schema reporting.
> - **Dependencies**:
>   - Add `jsonschema` to `pyproject.toml`.
> - **Tests**:
>   - Comprehensive unit tests in `tests/unit_tests/test_schema_validator.py` covering defaults, nullable types, arrays/objects, validation, and error handling.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit dafc07d7232d9c0eb9ca6e2797e238ca71fe3774. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: skyvern/forge/sdk/api/llm/schema_validator.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+        LOG.debug("Schema is a string, treating as permissive", path=path, schema=schema)
+        return data
+
+    if isinstance(schema, list):
+        LOG.debug("Schema is a list, treating as permissive", path=path)
+        return data
+
+    schema_type = schema.get("type")
+
+    if schema_type == "object" or "properties" in schema:
```
Comment (marcmuon) [inline skyvern/forge/sdk/api/llm/schema_validator.py:78]: only question on PR is does this still work for nullable types where schema is like `["object", "null"]`. i think for those we'd skip this validation (but we'd still wanna validate)? lmk what you think @mohamedmamdouh22

File: skyvern/forge/sdk/api/llm/schema_validator.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        return []
+    elif schema_type_str == "object":
+        return {}
+    elif schema_type_str == "null":
+        return None
+    else:
+        return None
+
+
+def fill_missing_fields(data: Any, schema: dict[str, Any] | list | str | None, path: str = "root") -> Any:
```
Comment (suchintan) [inline skyvern/forge/sdk/api/llm/schema_validator.py:49]: Question: is tehre a reason why we wouldn't want to use jsonschema to validate the input schema (https://python-jsonschema.readthedocs.io/en/stable/)?

If the schema is invalid.. maybe we can update the FE to notify the user that the schema is invalid for (XYZ reasons), and we can't guarantee type correctness

IF the schema is valid... chatGPT suggested this might work:

--- CHAT GPT GENERATED OUTPUT ---
2. Filling defaults from the schema (like your fill_missing_fields)

The JSON Schema spec does not require validators to apply default values automatically, so jsonschema doesn’t do that out of the box. But there’s a well-known pattern to extend the validator to write defaults into the instance while validating.

Here’s a compact version that replaces your custom recursion:
```
from typing import Any
import jsonschema
from jsonschema import Draft202012Validator

def extend_with_default(validator_class):
    """
    Return a validator class that also sets default values on the instance.
    """
    validate_properties = validator_class.VALIDATORS["properties"]

    def set_defaults(validator, properties, instance, schema):
        # Only touch dict-like instances
        if not isinstance(instance, dict):
            yield from validate_properties(validator, properties, instance, schema)
            return

        for prop, subschema in properties.items():
            if "default" in subschema and prop not in instance:
                instance[prop] = subschema["default"]

        # Continue with normal validation
        yield from validate_properties(validator, properties, instance, schema)

    return jsonschema.validators.extend(
        validator_class,
        {"properties": set_defaults},
    )

DefaultingValidator = extend_with_default(Draft202012Validator)
```

Usage:
```
def validate_and_fill_extraction_result(
    extraction_result: dict[str, Any],
    schema: dict[str, Any] | None,
) -> dict[str, Any]:
    if schema is None:
        LOG.debug("No schema provided, returning extraction result as-is")
        return extraction_result

    LOG.info("Validating and filling extraction result against schema")

    try:
        # Make a copy if you don't want in-place mutation
        result = dict(extraction_result)
        validator = DefaultingValidator(schema)
        validator.validate(result)
        LOG.info("Successfully validated and filled extraction result")
        return result
    except jsonschema.ValidationError as e:
        LOG.error(
            "Failed to validate and fill extraction result",
            error=str(e),
            path=list(e.path),
            schema_path=list(e.schema_path),
            exc_info=True,
        )
        # Decide your failure behavior here
        return extraction_result
```

File: skyvern/forge/sdk/api/llm/schema_validator.py (hunk 1 above)
Comment (mohamedmamdouh22) [reply skyvern/forge/sdk/api/llm/schema_validator.py:78]: Good catch @marcmuon , Let me know what about this handle 
```
    if isinstance(schema_type, list) and "null" in schema_type and data is None:
        LOG.debug("Data is None and schema allows null type, keeping as None", path=path)
        return None


    if isinstance(schema_type, list):
        non_null_types = [t for t in schema_type if t != "null"]
        if len(non_null_types) == 1:
            schema_type = non_null_types[0]
        elif len(non_null_types) > 1:
            LOG.warning(
                "Multiple non-null types in schema, using first one",
                path=path,
                types=non_null_types,
            )
            schema_type = non_null_types[0]
```

File: tests/unit_tests/test_schema_validator.py [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+            "properties": {
+                "name": {"type": "string"},
+                "age": {"type": "integer"},
+            },
+            "required": ["name", "age"],
+        }
+
+        # When data is null, it should remain null (valid for nullable type)
+        result = fill_missing_fields(None, schema)
+        assert result is None
```
Comment (mohamedmamdouh22) [inline tests/unit_tests/test_schema_validator.py:275]: @marcmuon this what you meant right?

Comment (marcmuon) [reply tests/unit_tests/test_schema_validator.py:275]: yes ty!

File: skyvern/forge/sdk/api/llm/schema_validator.py (hunk 1 above)
Comment (marcmuon) [reply skyvern/forge/sdk/api/llm/schema_validator.py:78]: code in latest commit looks good to me

File: skyvern/forge/sdk/api/llm/schema_validator.py [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+
+    if schema_type == "object" or "properties" in schema:
+        # If data is None and schema allows null, keep it as None
+        if data is None and is_nullable:
+            LOG.debug("Data is None and schema allows null, keeping as None", path=path)
+            return None
+
+        if not isinstance(data, dict):
+            LOG.warning(
+                "Expected object but got different type, creating empty object",
```
Comment (suchintan) [inline skyvern/forge/sdk/api/llm/schema_validator.py:86]: Can we update this message? `empty object` needs a bit more context on why you're creating an empty object

File: skyvern/forge/sdk/api/llm/schema_validator.py [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+                    LOG.debug("Skipping optional missing field", path=field_path)
+                    continue
+
+            data[field_name] = fill_missing_fields(data[field_name], field_schema, field_path)
+
+        return data
+
+    if schema_type == "array":
+        # If data is None and schema allows null, keep it as None
+        if data is None and is_nullable:
```
Comment (suchintan) [inline skyvern/forge/sdk/api/llm/schema_validator.py:119]: this is repeated

File: skyvern/forge/sdk/api/llm/schema_validator.py [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+    """
+    if schema is None or isinstance(schema, (str, list)):
+        return
+
+    try:
+        Draft202012Validator.check_schema(schema)
+    except SchemaError as e:
+        error_message = f"Invalid JSON schema: {str(e)}"
+        LOG.error("Invalid JSON schema", error=str(e), schema=schema)
+        raise InvalidSchemaError(error_message, [str(e)])
```
Comment (suchintan) [inline skyvern/forge/sdk/api/llm/schema_validator.py]: You probably don't want to fail a run if the schema is invalid - LLMs are surprisingly resilient to this

Comment (suchintan) [reply skyvern/forge/sdk/api/llm/schema_validator.py]: You should return `true` or `false` to indicate schema valid or not

Comment (suchintan) [reply skyvern/forge/sdk/api/llm/schema_validator.py]: And if the schema is invalid, just return `data` as-is. No need to make transformations

File: skyvern/forge/sdk/api/llm/schema_validator.py (hunk 4 above)
Comment (mohamedmamdouh22) [reply skyvern/forge/sdk/api/llm/schema_validator.py:86]:
```suggestion
                "Schema expects object type but received incompatible type, creating empty object to satisfy schema requirements and continue validation",
```

File: skyvern/forge/sdk/api/llm/schema_validator.py (hunk 6 above)
Comment (mohamedmamdouh22) [reply skyvern/forge/sdk/api/llm/schema_validator.py]: so we do not need to raise exception for the invalide schema?
```
    if schema is None or isinstance(schema, (str, list)):
        return True

    try:
        Draft202012Validator.check_schema(schema)
        return True
    except SchemaError as e:
        LOG.warning("Invalid JSON schema, will return data as-is", error=str(e), schema=schema)
        return False
   ```

File: skyvern/forge/sdk/api/llm/schema_validator.py (hunk 5 above)
Comment (mohamedmamdouh22) [reply skyvern/forge/sdk/api/llm/schema_validator.py:119]: sorry i did not get you
Objects and arrays both need to handle nullable types
The logic is identical but applies to different schema types in lines 80-84 for object and lines 120-124 for lists

File: skyvern/forge/sdk/api/llm/schema_validator.py (hunk 6 above)
Comment (LawyZheng) [reply skyvern/forge/sdk/api/llm/schema_validator.py]: You can raise executions if you want. But what @suchintan  wanted to say is — if the schema is invalid, we can trust LLM’s response, and return the json response directly,.

File: skyvern/forge/sdk/api/llm/schema_validator.py [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+
+_TYPE_DEFAULT_FACTORIES: dict[str, Any] = {
+    "string": lambda: "",
+    "number": lambda: 0,
+    "integer": lambda: 0,
+    "boolean": lambda: False,
+    "array": list,
+    "object": dict,
+    "null": lambda: None,
+}
```
Comment (LawyZheng) [inline skyvern/forge/sdk/api/llm/schema_validator.py:18]: Nit: is it possible that user will use the upper case format like `STRING` for the schema type?

File: skyvern/forge/sdk/api/llm/schema_validator.py (hunk 5 above)
Comment (suchintan) [reply skyvern/forge/sdk/api/llm/schema_validator.py:119]: OK Sounds good

File: skyvern/forge/sdk/api/llm/schema_validator.py (hunk 6 above)
Comment (suchintan) [reply skyvern/forge/sdk/api/llm/schema_validator.py]: Exactly

Let's say a customer has defined an invalid schema in production. This change would now start breaking that automation with an exception, whereas it worked well enough before

Comment (mohamedmamdouh22) [reply skyvern/forge/sdk/api/llm/schema_validator.py]: Got you guys, Ok I will account for that

File: skyvern/forge/sdk/api/llm/schema_validator.py (hunk 7 above)
Comment (mohamedmamdouh22) [reply skyvern/forge/sdk/api/llm/schema_validator.py:18]: Addressed and added a test for this

File: skyvern/forge/sdk/api/llm/schema_validator.py (hunk 6 above)
Comment (mohamedmamdouh22) [reply skyvern/forge/sdk/api/llm/schema_validator.py]: ok this is addressed here
```
    if not validate_schema(schema):
        LOG.info("Schema is invalid, returning extraction result as-is without transformations")
        return extraction_result
     ```

**Comment by LawyZheng** [comment]
Thanks for the contribution
